pkg/compactor/executor.go

```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package compactor
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/ring"
+	"github.com/oklog/ulid/v2"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
+	"google.golang.org/grpc"
+
+	"github.com/grafana/mimir/pkg/compactor/scheduler/schedulerpb"
+	"github.com/grafana/mimir/pkg/util"
+)
+
+// CompactionExecutor defines how compaction work is executed.
+type CompactionExecutor interface {
+	Run(ctx context.Context, compactor *MultitenantCompactor) error
+	CreateShardingStrategy(enabledTenants, disabledTenants []string, ring *ring.Ring, ringLifecycler *ring.BasicLifecycler, cfgProvider ConfigProvider) shardingStrategy
+	Stop() error
+}
+
+// StandaloneExecutor runs compaction on a timer without external scheduling.
+type StandaloneExecutor struct{}
+
+func (e *StandaloneExecutor) Run(ctx context.Context, c *MultitenantCompactor) error {
+
+	c.compactUsers(ctx)
+
+	ticker := time.NewTicker(util.DurationWithJitter(c.compactorCfg.CompactionInterval, 0.05))
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			c.compactUsers(ctx)
+		case <-ctx.Done():
+			return nil
+		case err := <-c.ringSubservicesWatcher.Chan():
+			return errors.Wrap(err, "compactor subservice failed")
+		}
+	}
+}
+
+func (e *StandaloneExecutor) Stop() error {
+	return nil
+}
+
+func (e *StandaloneExecutor) CreateShardingStrategy(enabledTenants, disabledTenants []string, ring *ring.Ring, ringLifecycler *ring.BasicLifecycler, cfgProvider ConfigProvider) shardingStrategy {
+	allowedTenants := util.NewAllowList(enabledTenants, disabledTenants)
+	return newSplitAndMergeShardingStrategy(allowedTenants, ring, ringLifecycler, cfgProvider)
+}
+
+// SchedulerExecutor requests compaction jobs from an external scheduler.
+type SchedulerExecutor struct {
```

Comment:
Anything in this file that is only used in the `compactor` package can (should) be unexported.